op-conductor/consensus/raft.go

```diff
@@ -0,0 +1,226 @@
+	if err := rc.r.AddVoter(raft.ServerID(id), raft.ServerAddress(addr), 0, defaultTimeout).Error(); err != nil {
+		rc.log.Error("failed to add voter", "id", id, "addr", addr, "err", err)
+		return err
+	}
+	return nil
+}
+
+// DemoteVoter implements Consensus, it tries to demote a voting member into a non-voting member in the cluster.
+func (rc *RaftConsensus) DemoteVoter(id string) error {
+	if err := rc.r.DemoteVoter(raft.ServerID(id), 0, defaultTimeout).Error(); err != nil {
+		rc.log.Error("failed to demote voter", "id", id, "err", err)
+		return err
+	}
+	return nil
+}
+
+// Leader implements Consensus, it returns true if it is the leader of the cluster.
+func (rc *RaftConsensus) Leader() bool {
+	_, id := rc.r.LeaderWithID()
+	return id == rc.serverID
+}
+
+// LeaderCh implements Consensus, it returns a channel that will be notified when leadership status changes (true = leader, false = follower).
+func (rc *RaftConsensus) LeaderCh() <-chan bool {
+	return rc.r.LeaderCh()
+}
+
+// RemoveServer implements Consensus, it tries to remove a member (both voter or non-voter) from the cluster, if leader is being removed, it will cause a new leader election.
+func (rc *RaftConsensus) RemoveServer(id string) error {
+	if err := rc.r.RemoveServer(raft.ServerID(id), 0, defaultTimeout).Error(); err != nil {
+		rc.log.Error("failed to remove voter", "id", id, "err", err)
+		return err
+	}
+	return nil
+}
+
+// ServerID implements Consensus, it returns the server ID of the current server.
+func (rc *RaftConsensus) ServerID() string {
+	return string(rc.serverID)
+}
+
+// TransferLeader implements Consensus, it triggers leadership transfer to another member in the cluster.
+func (rc *RaftConsensus) TransferLeader() error {
+	if err := rc.r.LeadershipTransfer().Error(); err != nil {
+		// Expected error if not leader
+		if errors.Is(err, raft.ErrNotLeader) {
+			return nil
+		}
+
+		rc.log.Error("failed to transfer leadership", "err", err)
+		return err
+	}
+	return nil
+}
+
+// TransferLeaderTo implements Consensus, it triggers leadership transfer to a specific member in the cluster.
+func (rc *RaftConsensus) TransferLeaderTo(id string, addr string) error {
+	if err := rc.r.LeadershipTransferToServer(raft.ServerID(id), raft.ServerAddress(addr)).Error(); err != nil {
+		rc.log.Error("failed to transfer leadership to server", "id", id, "addr", addr, "err", err)
+		return err
+	}
+	return nil
+}
+
+// Shutdown implements Consensus, it shuts down the consensus protocol client.
+func (rc *RaftConsensus) Shutdown() error {
+	if err := rc.r.Shutdown().Error(); err != nil {
+		rc.log.Error("failed to shutdown raft", "err", err)
+		return err
+	}
+	return nil
+}
+
+// CommitUnsafePayload implements Consensus, it commits latest unsafe payload to the cluster FSM.
+func (rc *RaftConsensus) CommitUnsafePayload(payload eth.ExecutionPayload) error {
+	blockVersion := eth.BlockV1
+	if rc.rollupCfg.IsCanyon(uint64(payload.Timestamp)) {
```

Comment:
Is it strictly necessary to make `op-conductor` aware of hardforks? Wouldn't it be cleaner to treat the unsafe heads data coming from Geth as a black box?

Reply:
Yes, it's necessary, in order to start sequencer properly at the consensus block, we need:
1. know the block number and hash to make sure sequencer has the latest block
2. and if not, need to gossip the block to op-node from consensus, and we need to know if we should publish to `blocksV1` or `blocksV2` topic

Reply:
got it

Reply:
Dumb question: can we remove this code once canyon has passed on mainnet?

Reply:
Definitely can